### ISSUE TITLE
Proper usage of :public_key.cacerts_get/0

### ISSUE THREAD

**Issue by nherzing** [elixir-mint/mint#377]
The [documentation](https://github.com/elixir-mint/mint/blob/v1.4.2/lib/mint/http.ex#L100) shows that the result of `:public_keys.cacerts_get/0` can be used as
```elixir
Mint.connect(:https, host, port, transport_opts: [cacerts: :public_key.cacerts_get()])
```
(I'm assuming it should be `Mint.HTTP.connect`)

However, this raises the following for me
```elixir
** (FunctionClauseError) no function clause matching in :public_key.pkix_decode_cert/2

    The following arguments were given to :public_key.pkix_decode_cert/2:

        # 1
        {:cert,
         <<48, 130, 4, ...
        # full cert details elided

        # 2
        :plain

    (public_key 1.13.1) public_key.erl:516: :public_key.pkix_decode_cert/2
    (elixir 1.14.1) lib/enum.ex:1658: Enum."-map/2-lists^map/1-0-"/2
    (mint 1.4.2) lib/mint/core/transport/ssl.ex:572: Mint.Core.Transport.SSL.add_partial_chain_fun/1
    (mint 1.4.2) lib/mint/core/transport/ssl.ex:443: Mint.Core.Transport.SSL.add_verify_opts/2
    (mint 1.4.2) lib/mint/core/transport/ssl.ex:432: Mint.Core.Transport.SSL.ssl_opts/2
    (mint 1.4.2) lib/mint/core/transport/ssl.ex:328: Mint.Core.Transport.SSL.connect/4
    (mint 1.4.2) lib/mint/negotiate.ex:67: Mint.Negotiate.connect_negotiate/4
    iex:1: (file)
```

Reading the erlang [`:public_key` docs](https://www.erlang.org/doc/man/public_key.html), this seems expected to me. Mint is passing the result of `:public_key.cacerts_get/0` which is already decoded into `:public_key.pkix_decode_cert/2` which expects a binary. Deleting [line 572 of ssl.ex](https://github.com/elixir-mint/mint/blob/56c1734ea946a387377deeceaead33a900c4041e/lib/mint/core/transport/ssl.ex#L572) makes it function as expected but that would obviously break passing in encoded certs.

Is there some other existing way to make use of `:public_keys.cacerts_get/0` in Mint that I am missing? Or will Mint need to be updated to properly work with `:public_keys.cacerts_get/0`?

Elixir version: 1.14.1
Erlang version: 25.1.1
Mint version: 1.4.2

**Comment by whatyouhide** [comment]
I think we'll need to patch Mint, yep. Any chance you'd like to work on a PR? 🙃

**Comment by beligante** [comment]
Ohh, this was a good finding hahaha. 
I've merged an adapter to use mint on elixir-grpc (https://github.com/elixir-grpc/grpc/pull/272) and I notice that ssl connections stoped to work.

@nherzing are you working on this? Otherwise I can

**Comment by whatyouhide** [comment]
@beligante go ahead please, very appreciated! 💟 🙃

**Comment by beligante** [comment]
Hey @whatyouhide working on this now - sorry for the delay.

I could follow a few options here. I like your input. The `:public_key.cacerts_get/0` returns a list with a three position tuple.
`{:cert, encoded_cert_in_binary, #'OTPCertificate'{}}`

in the map function I thought in add a pattern match case for handling this list 

```elixir
  defp decode_cacerts(certs) do
    Enum.map(certs, fn 
      cert when is_binary(cert) ->  :public_key.pkix_decode_cert(cert, :plain)
      {:cert, cert, _otp_cert} when is_binary(cert) -> :public_key.pkix_decode_cert(cert, :plain)
    end)
  end
```

Second option is enforce that option to be a list of binaries for that option and raise in case something doesn't match. This way we could simply update the docs like:

```elixir
certs = :public_key.cacerts_get() |> Enum.map(fn {:cert, binary, _otp_cert} -> binary end)
Mint.connect(:https, host, port, transport_opts: [cacerts: :public_key.cacerts_get()])
```

Thoghts?

PS: @nherzing the above code could be an option for you to use the trusted erlang certs without code changes on mint

**Comment by nherzing** [comment]
@beligante Just seeing these comments now. I initially took a shot at making essentially the exact same change you are proposing. Testing with my local certs, I came across another case with a 4-tuple tagged with `:Certificate`. I also think in the case that we see a tuple, we can just pass the value through. I don't believe there's a reason to re-decode it. Having said that, I'm definitely out of my depth with Erlang's handling of ssl certs so take this all with a grain of salt.

```elixir
defp decode_cacerts(certs) do
  Enum.map(certs, fn
    {:Certificate, _, _ , _} = cert -> cert
    {:cert, _, _} = cert -> cert
    cert -> :public_key.pkix_decode_cert(cert, :plain)
  end)
end
```

I ended up just directly specifying the certfile I needed as a workaround.
```elixir
Mint.connect(:https, host, port, transport_opts: [cacertfile: Application.app_dir(:my_app) <> "/priv/internal_root.crt"])
```
